app/src/main/java/org/thoughtcrime/securesms/backup/FullBackupExporter.java

```diff
@@ -0,0 +1,611 @@
+import org.thoughtcrime.securesms.keyvalue.SignalStore;
+import org.thoughtcrime.securesms.profiles.AvatarHelper;
+import org.thoughtcrime.securesms.service.PendingRetryReceiptManager;
+import org.thoughtcrime.securesms.util.SetUtil;
+import org.thoughtcrime.securesms.util.Stopwatch;
+import org.thoughtcrime.securesms.util.TextSecurePreferences;
+import org.thoughtcrime.securesms.util.Util;
+import org.whispersystems.libsignal.kdf.HKDFv3;
+import org.whispersystems.libsignal.util.ByteUtil;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.Mac;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+public class FullBackupExporter extends FullBackupBase {
+
+  private static final String TAG = Log.tag(FullBackupExporter.class);
+
+  private static final Set<String> BLACKLISTED_TABLES = SetUtil.newHashSet(
+    SignedPreKeyDatabase.TABLE_NAME,
+    OneTimePreKeyDatabase.TABLE_NAME,
+    SessionDatabase.TABLE_NAME,
+    SearchDatabase.SMS_FTS_TABLE_NAME,
+    SearchDatabase.MMS_FTS_TABLE_NAME,
+    EmojiSearchDatabase.TABLE_NAME,
+    SenderKeyDatabase.TABLE_NAME,
+    SenderKeySharedDatabase.TABLE_NAME,
+    PendingRetryReceiptDatabase.TABLE_NAME,
+    AvatarPickerDatabase.TABLE_NAME
+  );
+
+  public static void export(@NonNull Context context,
+                            @NonNull AttachmentSecret attachmentSecret,
+                            @NonNull SQLiteDatabase input,
+                            @NonNull File output,
+                            @NonNull String passphrase,
+                            @NonNull BackupCancellationSignal cancellationSignal)
+      throws IOException
+  {
+    try (OutputStream outputStream = new FileOutputStream(output)) {
+      internalExport(context, attachmentSecret, input, outputStream, passphrase, true, cancellationSignal);
+    }
+  }
+
+  @RequiresApi(29)
+  public static void export(@NonNull Context context,
+                            @NonNull AttachmentSecret attachmentSecret,
+                            @NonNull SQLiteDatabase input,
+                            @NonNull DocumentFile output,
+                            @NonNull String passphrase,
+                            @NonNull BackupCancellationSignal cancellationSignal)
+      throws IOException
+  {
+    try (OutputStream outputStream = Objects.requireNonNull(context.getContentResolver().openOutputStream(output.getUri()))) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openOutputStream*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at line 116 evaluates to true. Ensure the object returned by *openOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).